File: vscode-extension/src/pathReviewConfig.ts
// vscode-extension/src/pathReviewConfig.ts
// -----------------------------------------------------------------------------
// Reads a repo's own .gto.yaml — per-path review rules (e.g. "skip the
// performance agent under scripts/**", "steer stricter review for
// payments/**"). Same file, same schema, and same trust treatment as the
// backend's ingestion/path_review_config.py — the VS Code extension already
// has full workspace filesystem access, so it loads the file directly and
// sends the parsed content along with the analysis request (see
// core.models.AnalysisRequest.path_review_config on the backend, which
// re-scans any free text here through governance.prompt_guard before it
// ever reaches an LLM prompt — never trust this file's content by itself).
//
// Malformed or missing .gto.yaml is never an error — it just means no
// path-scoped rules for this run, same failure mode as any other optional
// per-repo file (mirrors reportState.ts's .gto-ignore.json handling).

import * as vscode from 'vscode';
import * as yaml from 'js-yaml';

export interface PathReviewRule {
  match: string;
  agents?: string[] | null;
  user_instructions?: string;
  skip?: boolean;
}

export interface PathReviewConfig {
  version: number;
  paths: PathReviewRule[];
}

const CONFIG_FILE = '.gto.yaml';
const MAX_CONFIG_BYTES = 64_000; // defensive cap — this is a small hand-written file

function configUri(repoRoot: string): vscode.Uri {
  return vscode.Uri.joinPath(vscode.Uri.file(repoRoot), CONFIG_FILE);
}

function isValidRule(v: unknown): v is PathReviewRule {
  if (!v || typeof v !== 'object') return false;
  const r = v as Record<string, unknown>;
  return typeof r.match === 'string' && r.match.length > 0;
}

/** Parses already-read .gto.yaml text. Exported separately from
 * loadPathReviewConfig so it's testable without touching the filesystem. */
export function parsePathReviewConfig(raw: string): PathReviewConfig | undefined {
  if (!raw || raw.length > MAX_CONFIG_BYTES) return undefined;
  try {
    const data = yaml.load(raw);
    if (!data || typeof data !== 'object') return undefined;
    const d = data as Record<string, unknown>;
    const rawPaths = Array.isArray(d.paths) ? d.paths : [];
    const paths = rawPaths.filter(isValidRule);
    if (!paths.length) return undefined;
    return { version: typeof d.version === 'number' ? d.version : 1, paths };
  } catch {
    return undefined; // malformed YAML — proceed without path-scoped rules
  }
}

export async function loadPathReviewConfig(repoRoot: string): Promise<PathReviewConfig | undefined> {
  try {
    const bytes = await vscode.workspace.fs.readFile(configUri(repoRoot));
    return parsePathReviewConfig(Buffer.from(bytes).toString('utf8'));
  } catch {
    return undefined; // file doesn't exist — the common case, not an error
  }
}

File: intellij-plugin/src/main/kotlin/com/uob/gto/PathReviewConfig.kt
package com.uob.gto

import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import org.yaml.snakeyaml.Yaml
import java.io.File

/**
 * Reads a repo's own .gto.yaml — per-path review rules. Kotlin port of
 * vscode-extension/src/pathReviewConfig.ts (same file, same schema, same
 * trust treatment: the backend re-scans any free text here through
 * governance.prompt_guard before it ever reaches an LLM prompt — this
 * client-side parse is never itself a trust boundary). Malformed or missing
 * .gto.yaml is never an error — same failure mode as GtoReportState's
 * .gto-ignore.json handling.
 */
object PathReviewConfig {

    private const val CONFIG_FILE = ".gto.yaml"
    private const val MAX_CONFIG_BYTES = 64_000

    data class Rule(val match: String, val agents: List<String>?, val userInstructions: String?, val skip: Boolean)
    data class Config(val version: Int, val paths: List<Rule>)

    @Suppress("UNCHECKED_CAST")
    private fun isValidRule(v: Any?): Boolean {
        val m = v as? Map<String, Any?> ?: return false
        return (m["match"] as? String)?.isNotEmpty() == true
    }

    /** Parses already-read .gto.yaml text — separated from [load] so it's
     * testable without touching the filesystem, same as parsePathReviewConfig
     * in the VS Code source. */
    @Suppress("UNCHECKED_CAST")
    fun parse(raw: String): Config? {
        if (raw.isEmpty() || raw.length > MAX_CONFIG_BYTES) return null
        return try {
            val data = Yaml().load<Any?>(raw) as? Map<String, Any?> ?: return null
            val rawPaths = data["paths"] as? List<Any?> ?: emptyList()
            val paths = rawPaths.filter { isValidRule(it) }.map {
                val m = it as Map<String, Any?>
                Rule(
                    match = m["match"] as String,
                    agents = (m["agents"] as? List<Any?>)?.map { a -> a.toString() },
                    userInstructions = m["user_instructions"] as? String,
                    skip = m["skip"] as? Boolean ?: false,
                )
            }
            if (paths.isEmpty()) null else Config(version = (data["version"] as? Number)?.toInt() ?: 1, paths = paths)
        } catch (e: Exception) {
            null // malformed YAML — proceed without path-scoped rules
        }
    }

    fun load(repoRoot: String): Config? {
        val file = File(repoRoot, CONFIG_FILE)
        if (!file.exists()) return null
        return try {
            parse(file.readText())
        } catch (e: Exception) {
            null
        }
    }

    /** Serializes to the JSON shape the backend's AnalysisRequest.
     * path_review_config field expects — same as what apiClient.ts sends
     * (a plain object, not string-keyed differently). */
    fun toJson(config: Config): JsonObject = buildJsonObject {
        put("version", config.version)
        put("paths", buildJsonArray {
            config.paths.forEach { rule ->
                add(buildJsonObject {
                    put("match", rule.match)
                    if (rule.agents != null) {
                        put("agents", buildJsonArray { rule.agents.forEach { add(JsonPrimitive(it)) } })
                    }
                    if (rule.userInstructions != null) put("user_instructions", rule.userInstructions)
                    put("skip", rule.skip)
                })
            }
        })
    }
}
